provider/ec2/environ.go

```diff
@@ -668,12 +687,57 @@ func (e *environ) StartInstance(
 		// Tags currently not supported by EC2
 		AvailabilityZone: &instAZ,
 	}
+
+	if err := e.attachInstanceProfile(ctx, callback, inst, args.Constraints); err != nil {
+		return nil, err
+	}
+
 	return &environs.StartInstanceResult{
 		Instance: inst,
 		Hardware: &hc,
 	}, nil
 }
 
+// attachInstanceProfile assesses if an instance profile needs to be attached
+// to an instance based on it's constraints. If the instance constraints do not
+// specify an instance role then this func returns silently.
+func (e *environ) attachInstanceProfile(
```

Comment:
We tend to call methods like this "maybeAttachInstanceProfile" as it only does it if required by the constraints